### PR TITLE
Header line

### PR DESCRIPTION
**Pull request by zamblauskas**
Added config option `contain_header` (default is `false`), if set to `true` it will use first line of csv input as column names.
Added config option `stream_identity` which allows to differentiate streams (e.g. csv files) and use different column names (and|or order) for each file.
We use it in production for couple months now - works nicely.
Tested against `logstash-1.5.0`.

### CODE REVIEW

**Comment by zamblauskas** [comment]
I had signed the CLA, but my email was not made public on GitHub profile. I suspect this is the reason for `CLA — Some commit authors have not signed the CLA or are not members of...`
I have now made the email public, is it possible to remove this CLA error ?

File: lib/logstash/filters/csv.rb [hunk 1]
```diff
@@ -22,6 +22,18 @@ class LogStash::Filters::CSV < LogStash::Filters::Base
   # (e.g. "user_defined_1", "user_defined_2", "column3", "column4", etc.)
   config :columns, :validate => :array, :default => []
 
+  # Define if CSV data contain header line. If `contain_header` is `true`,
+  # first line will be parsed as column names. In the case that there are more columns
+  # in the data than parsed from the header line, specified `columns` will be used.
+  # In the case `columns` are not specified or data contain more columns than
+  # there are `columns` specified, auto-numbered columns will be used.
+  config :contain_header, :validate => :boolean, :default => false
```
Comment (magnusbaeck) [inline lib/logstash/filters/csv.rb]: Perhaps contain**s**_header would be a better name?

**Comment by magnusbaeck** [comment]
What if Logstash is restarted in the middle of the processing of a file? I don't see it persisting the captured column names so it seems they would be lost.

**Comment by jordansissel** [comment]
@zamblauskas the CLA check uses the git commit author on the commits, not your github email address. Further, this check is only done per _commit_ not periodically. I've run my script to re-check everything and things are looking ok on CLA :) Thanks!

**Comment by jordansissel** [comment]
> What if Logstash is restarted in the middle of the processing of a file? 

I agree with this concern. We have plans to make a seekable file input that will allow something like a `csv` codec to remember where in a file it last read, but also be able to re-learn the headers for a given CSV from the beginning. This isn't possible today, though.

File: lib/logstash/filters/csv.rb (hunk 1 above)
Comment (zamblauskas) [inline lib/logstash/filters/csv.rb]: > Perhaps contains_header would be a better name?

yep, contain**s**_header it will be, thanks for noting this

**Comment by zamblauskas** [comment]
> What if Logstash is restarted in the middle of the processing of a file?

Not an issue in our use case - we process a large number of small files, each single time only. 
But might be a problem for other users - agree.
Will take a look how sincedb is used in logstash and how easy it would be to save the header.

**Comment by zamblauskas** [comment]
1) I expected `sincedb` to be some kind of a self-contained full-blown db, turns out it's just a simple file :) But it's possible to save the headers in a similar manner.
2) After little bit more thinking about saving headers in a db/file, I think it's not possible to do it properly within a filter.
Main obstacle is rotating files - the filter wouldn't know that a rotation happened and that it needs to re-read the headers for a given file.
That `seekable file input` @jordansissel mentioned is required in this case, hence leaving this PR as it is for now.

**Comment by steverowe** [comment]
This is an important feature.  Is there anything I can do to help out getting this request integrated?

**Comment by markwalkom** [comment]
Wanted to chase this one up @jordansissel @magnusbaeck; isn't the sincedb best handled by the input itself? Why do we care if we drop a line once it has passed that point?

To my (un-ruby) mind, this is the same as using something like `  if "_dateparsefailure" in [tags] { drop {} }` if you have a date column and you want to easily drop the header from the output. If I wanted to reprocess the file then I know the input (assuming `file`) will handle it.

**Comment by colinmollenhour** [comment]
Just FYI it looks like the PR that I submitted that got merged has a little overlap with this one (just the contains_header part).. See #36.

**Comment by guyboertje** [comment]
Closing as this approach is no longer suitable.

There are too many assumptions:
- a file path being always available (stream identity).
- not file tailing.
- using the file input.

**Pull request closed without merging**